File: backend/internal/service/finance.go
package service

import (
	"errors"
	"fmt"
	"strings"

	"monman-backend/internal/models"
	"monman-backend/internal/repository"

	"github.com/google/uuid"
)

type validationError struct{ msg string }

func (e validationError) Error() string { return e.msg }

// IsValidation reports client-side input errors suitable for HTTP 400.
func IsValidation(err error) bool {
	_, ok := err.(validationError)
	return ok
}

// FinanceService aggregates dashboard, lists, account/category/budget payloads, and creates.
type FinanceService struct {
	txRepo  *repository.TransactionRepository
	accRepo *repository.AccountRepository
	catRepo *repository.CategoryRepository
	budRepo *repository.BudgetRepository
}

func NewFinanceService(
	txRepo *repository.TransactionRepository,
	accRepo *repository.AccountRepository,
	catRepo *repository.CategoryRepository,
	budRepo *repository.BudgetRepository,
) *FinanceService {
	return &FinanceService{
		txRepo:  txRepo,
		accRepo: accRepo,
		catRepo: catRepo,
		budRepo: budRepo,
	}
}

// Dashboard returns balance, monthly net, and last N transactions.
func (s *FinanceService) Dashboard(userID uuid.UUID, recentLimit int) (*models.DashboardPayload, error) {
	if recentLimit <= 0 || recentLimit > 50 {
		recentLimit = 8
	}
	total, err := s.txRepo.SumAccountBalance(userID)
	if err != nil {
		return nil, fmt.Errorf("dashboard balance: %w", err)
	}
	monthlyNet, err := s.txRepo.SumAmountForCalendarMonth(userID)
	if err != nil {
		return nil, fmt.Errorf("dashboard monthly: %w", err)
	}
	recent, err := s.txRepo.ListForUser(userID, recentLimit, 0)
	if err != nil {
		return nil, fmt.Errorf("dashboard recent: %w", err)
	}
	return &models.DashboardPayload{
		TotalBalanceCents:  total,
		MonthlyNetCents:    monthlyNet,
		RecentTransactions: recent,
	}, nil
}

// ListTransactions returns transactions with current-month income/expense totals.
func (s *FinanceService) ListTransactions(userID uuid.UUID, limit, offset int) (*models.TransactionListPayload, error) {
	list, err := s.txRepo.ListForUser(userID, limit, offset)
	if err != nil {
		return nil, err
	}
	inc, exp, err := s.txRepo.MonthIncomeExpenseCents(userID)
	if err != nil {
		return nil, err
	}
	return &models.TransactionListPayload{
		Transactions:      list,
		MonthIncomeCents:  inc,
		MonthExpenseCents: exp,
	}, nil
}

func (s *FinanceService) ListAccounts(userID uuid.UUID) ([]models.AccountSummary, error) {
	list, err := s.accRepo.ListActiveByUser(userID)
	if err != nil {
		return nil, err
	}
	if len(list) == 0 {
		if err := s.accRepo.EnsureDefaultCashWallet(userID); err != nil {
			return nil, err
		}
		list, err = s.accRepo.ListActiveByUser(userID)
		if err != nil {
			return nil, err
		}
	}
	return list, nil
}

func (s *FinanceService) ListCategories(userID uuid.UUID, typeFilter string) ([]models.CategorySummary, error) {
	return s.catRepo.ListActiveForUser(userID, typeFilter)
}

// CreateAccount inserts an additional wallet/account row for the user.
func (s *FinanceService) CreateAccount(userID uuid.UUID, req *models.CreateAccountRequest) (uuid.UUID, error) {
	name := strings.TrimSpace(req.Name)
	if name == "" {
		return uuid.Nil, validationError{"name is required"}
	}
	if len(name) > 120 {
		return uuid.Nil, validationError{"name is too long"}
	}
	at := strings.ToLower(strings.TrimSpace(req.AccountType))
	if at == "" {
		at = "cash"
	}
	switch at {
	case "bank", "credit_card", "cash", "investment", "ewallet":
	default:
		return uuid.Nil, validationError{"account_type must be bank, credit_card, cash, investment, or ewallet"}
	}
	col := strings.TrimSpace(req.Color)
	if col == "" {
		col = "#3B82F6"
	}
	return s.accRepo.Create(userID, name, at, col)
}

func (s *FinanceService) ListBudgets(userID uuid.UUID) (*models.BudgetsPayload, error) {
	cards, err := s.budRepo.ListBudgetCardsPayload(userID)
	if err != nil {
		return nil, err
	}
	return &models.BudgetsPayload{Budgets: cards}, nil
}

func (s *FinanceService) CreateBudget(userID uuid.UUID, req *models.CreateBudgetRequest) (uuid.UUID, error) {
	if req.AllocatedAmount <= 0 {
		return uuid.Nil, validationError{"allocated_amount must be positive"}
	}
	if strings.TrimSpace(req.Name) == "" {
		return uuid.Nil, validationError{"name is required"}
	}
	switch req.BudgetPeriod {
	case "weekly", "monthly", "yearly":
	default:
		return uuid.Nil, validationError{"invalid budget_period"}
	}
	if len(req.PeriodStartDate) < 8 || len(req.PeriodEndDate) < 8 {
		return uuid.Nil, validationError{"period_start_date and period_end_date are required (YYYY-MM-DD)"}
	}
	ctype, catOK, err := s.catRepo.CategoryOwnedOrSystem(req.CategoryID, userID)
	if err != nil {
		return uuid.Nil, err
	}
	if !catOK {
		return uuid.Nil, validationError{"category not found"}
	}
	if ctype != "expense" {
		return uuid.Nil, validationError{"budget category must be expense type"}
	}
	icon := "📱"
	if req.Icon != nil && strings.TrimSpace(*req.Icon) != "" {
		icon = strings.TrimSpace(*req.Icon)
	}
	color := "blue"
	if req.Color != nil && strings.TrimSpace(*req.Color) != "" {
		color = strings.TrimSpace(*req.Color)
	}
	sort := 0
	if req.SortOrder != nil {
		sort = *req.SortOrder
	}
	alert := 80
	if req.AlertPercentage != nil && *req.AlertPercentage > 0 && *req.AlertPercentage <= 100 {
		alert = *req.AlertPercentage
	}
	if len(req.CommonPurchases) > 80 {
		return uuid.Nil, validationError{"at most 80 common_purchases allowed"}
	}
	for _, cp := range req.CommonPurchases {
		if strings.TrimSpace(cp.Item) == "" {
			return uuid.Nil, validationError{"each common purchase needs item when listed"}
		}
		if cp.EstimatedAmount <= 0 {
			return uuid.Nil, validationError{"estimated_amount must be positive for common purchases"}
		}
	}
	id, err := s.budRepo.Create(userID, req.CategoryID, strings.TrimSpace(req.Name), req.AllocatedAmount,
		req.BudgetPeriod, req.PeriodStartDate, req.PeriodEndDate, req.ResponsiblePerson, alert,
		icon, color, sort, req.CommonPurchases)
	if err != nil {
		if strings.Contains(err.Error(), "UNIQUE constraint failed") || strings.Contains(err.Error(), "unique constraint") {
			return uuid.Nil, validationError{"a budget with this name already exists"}
		}
		return uuid.Nil, err
	}
	return id, nil
}

// AppendCommonPurchases inserts pembelian umum presets for an existing owned budget.
func (s *FinanceService) AppendCommonPurchases(userID uuid.UUID, budgetID uuid.UUID, req *models.AppendCommonPurchasesRequest) error {
	if req == nil || len(req.Purchases) == 0 {
		return validationError{"purchases required"}
	}
	if len(req.Purchases) > 80 {
		return validationError{"at most 80 purchases allowed"}
	}
	for _, cp := range req.Purchases {
		if strings.TrimSpace(cp.Item) == "" {
			return validationError{"each purchase needs item"}
		}
		if cp.EstimatedAmount <= 0 {
			return validationError{"estimated_amount must be positive"}
		}
	}
	err := s.budRepo.AppendCommonPurchases(budgetID, userID, req.Purchases)
	if errors.Is(err, repository.ErrBudgetNotOwned) {
		return validationError{"budget not found"}
	}
	return err
}

// CreateTransaction creates an income/expense posting and optionally links an expense line to one budget bucket.
func (s *FinanceService) CreateTransaction(userID uuid.UUID, req *models.CreateTransactionRequest) (uuid.UUID, error) {
	if req.MagnitudeAmountCents <= 0 {
		return uuid.Nil, validationError{"magnitude_amount_cents must be positive"}
	}
	if strings.TrimSpace(req.Description) == "" {
		return uuid.Nil, validationError{"description is required"}
	}
	if req.TransactionDate == "" {
		return uuid.Nil, validationError{"transaction_date is required"}
	}
	txType := strings.ToLower(strings.TrimSpace(req.TransactionType))
	if txType != "income" && txType != "expense" {
		return uuid.Nil, validationError{"transaction_type must be income or expense"}
	}

	ok, err := s.accRepo.AccountBelongs(req.AccountID, userID)
	if err != nil {
		return uuid.Nil, err
	}
	if !ok {
		return uuid.Nil, validationError{"account not found"}
	}

	var effectiveCategory uuid.UUID
	if req.BudgetID != nil {
		if txType != "expense" {
			return uuid.Nil, validationError{"budget_id can only be set for expense"}
		}
		bcat, okb, err := s.budRepo.BudgetBelongs(*req.BudgetID, userID)
		if err != nil {
			return uuid.Nil, err
		}
		if !okb {
			return uuid.Nil, validationError{"budget not found"}
		}
		effectiveCategory = bcat
		if req.CategoryID != nil && *req.CategoryID != effectiveCategory {
			return uuid.Nil, validationError{"category_id must match the budget category"}
		}
	} else {
		if req.CategoryID == nil {
			return uuid.Nil, validationError{"category_id is required"}
		}
		effectiveCategory = *req.CategoryID
	}

	ctype, ok, err := s.catRepo.CategoryOwnedOrSystem(effectiveCategory, userID)
	if err != nil {
		return uuid.Nil, err
	}
	if !ok {
		return uuid.Nil, validationError{"category not found"}
	}
	want := "expense"
	if txType == "income" {
		want = "income"
	}
	if ctype != want {
		return uuid.Nil, validationError{"category type does not match transaction_type"}
	}

	var signed int64
	switch txType {
	case "income":
		signed = req.MagnitudeAmountCents
	case "expense":
		signed = -req.MagnitudeAmountCents
	}

	var link *repository.BudgetLinkParams
	if req.BudgetID != nil && txType == "expense" {
		item := strings.TrimSpace(req.Description)
		if req.Item != nil && strings.TrimSpace(*req.Item) != "" {
			item = strings.TrimSpace(*req.Item)
		}
		link = &repository.BudgetLinkParams{
			BudgetID: *req.BudgetID,
			Item:     item,
			UserID:   userID,
		}
		if req.Quantity != nil {
			q := strings.TrimSpace(*req.Quantity)
			if q != "" {
				link.Quantity = &q
			}
		}
		if req.Store != nil {
			st := strings.TrimSpace(*req.Store)
			if st != "" {
				link.Store = &st
			}
		}
		mag := req.MagnitudeAmountCents
		link.UnitPrice = &mag
	}

	desc := strings.TrimSpace(req.Description)
	id, err := s.txRepo.Create(userID, req.AccountID, effectiveCategory, signed, desc,
		txType, req.TransactionDate, req.LocationName, link)
	if err != nil {
		return uuid.Nil, err
	}
	return id, nil
}

File: frontend/src/api/finance.ts
/**
 * Authenticated finance API — amounts are in cents (signed for amount).
 */

import {
  authenticatedRequest,
  type ApiResponse,
} from './client'

export interface TransactionApiRow {
  id: string
  date: string
  description: string
  category: string
  amount: number
  account: string
}

export interface DashboardPayload {
  total_balance_cents: number
  monthly_net_cents: number
  recent_transactions: TransactionApiRow[]
}

export interface TransactionsListPayload {
  transactions: TransactionApiRow[]
  month_income_cents: number
  month_expense_cents: number
}

function unwrapData<T>(
  _path: string,
  response: ApiResponse<T>,
  label: string
): T {
  if (response.status !== 'success' || response.data === undefined) {
    throw new Error(
      typeof response.error === 'string'
        ? response.error
        : `${label} failed`
    )
  }
  return response.data
}

/** GET /api/dashboard */
export async function getDashboard(recentLimit?: number): Promise<DashboardPayload> {
  let path = '/api/dashboard'
  if (recentLimit != null) {
    path += `?recent=${recentLimit}`
  }
  const res = await authenticatedRequest<
    ApiResponse<DashboardPayload>
  >(path)
  return unwrapData(path, res, 'Dashboard')
}

/** GET /api/transactions */
export async function getTransactions(opts?: {
  limit?: number
  offset?: number
}): Promise<TransactionsListPayload> {
  const params = new URLSearchParams()
  if (opts?.limit != null) params.set('limit', String(opts.limit))
  if (opts?.offset != null) params.set('offset', String(opts.offset))
  const q = params.toString()
  const path = q ? `/api/transactions?${q}` : '/api/transactions'
  const res = await authenticatedRequest<
    ApiResponse<TransactionsListPayload>
  >(path)
  return unwrapData(path, res, 'Transactions')
}

export interface AccountSummaryRow {
  id: string
  name: string
  account_type: string
  balance: number
  color: string
}

export interface CategorySummaryRow {
  id: string
  name: string
  category_type: string
  icon: string
  is_system: boolean
}

export interface BudgetCommonPurchaseRow {
  id: string
  item: string
  quantity?: string
  estimated_amount: number
  store?: string
  is_frequently_used: boolean
  sort_order: number
}

export interface BudgetLineItemRow {
  id: string
  transaction_id: string
  item: string
  quantity?: string
  store?: string
  amount_cents: number
  date: string
  transaction_description?: string
}

export interface BudgetCardRow {
  id: string
  name: string
  icon: string
  color: string
  category_display: string
  allocated_amount: number
  spent_amount: number
  budget_period: string
  period_start_date: string
  period_end_date: string
  category_id: string
  /** May be null from JSON when omitted or explicitly null — always coerce with ?? [] at use sites */
  common_purchases: BudgetCommonPurchaseRow[] | null
  line_items: BudgetLineItemRow[] | null
}

export interface BudgetsPayload {
  budgets: BudgetCardRow[]
}

export async function getAccounts(): Promise<AccountSummaryRow[]> {
  const res = await authenticatedRequest<
    ApiResponse<{ accounts: AccountSummaryRow[] }>
  >('/api/accounts')
  return unwrapData('/api/accounts', res, 'Accounts').accounts
}

export type AccountTypeOption =
  | 'bank'
  | 'credit_card'
  | 'cash'
  | 'investment'
  | 'ewallet'

export interface CreateAccountPayload {
  name: string
  account_type?: AccountTypeOption
  color?: string
}

/** POST /api/accounts */
export async function createAccount(body: CreateAccountPayload): Promise<{ id: string }> {
  const res = await authenticatedRequest<ApiResponse<{ id: string }>>('/api/accounts', {
    method: 'POST',
    body: JSON.stringify(body),
  })
  return unwrapData('/api/accounts', res, 'Create account')
}

export async function getCategories(opts?: {
  type?: 'income' | 'expense'
}): Promise<CategorySummaryRow[]> {
  const q = opts?.type ? `?type=${encodeURIComponent(opts.type)}` : ''
  const path = `/api/categories${q}`
  const res = await authenticatedRequest<
    ApiResponse<{ categories: CategorySummaryRow[] }>
  >(path)
  return unwrapData(path, res, 'Categories').categories
}

export async function getBudgets(): Promise<BudgetsPayload> {
  const res = await authenticatedRequest<ApiResponse<BudgetsPayload>>(
    '/api/budgets',
  )
  return unwrapData('/api/budgets', res, 'Budgets')
}

export interface CreateTransactionPayload {
  account_id: string
  category_id?: string
  magnitude_amount_cents: number
  description: string
  transaction_type: 'income' | 'expense'
  transaction_date: string
  location_name?: string
  budget_id?: string
  item?: string
  quantity?: string
  store?: string
}

/** POST /api/transactions */
export async function createTransaction(
  body: CreateTransactionPayload,
): Promise<{ id: string }> {
  const res = await authenticatedRequest<ApiResponse<{ id: string }>>(
    '/api/transactions',
    { method: 'POST', body: JSON.stringify(body) },
  )
  return unwrapData('/api/transactions', res, 'Create transaction')
}

export interface CreateCommonPurchasePayload {
  item: string
  quantity?: string
  /** cents */
  estimated_amount: number
  store?: string
  is_frequently_used?: boolean
  sort_order?: number
}

export interface CreateBudgetPayload {
  category_id: string
  name: string
  allocated_amount: number
  budget_period: 'weekly' | 'monthly' | 'yearly'
  period_start_date: string
  period_end_date: string
  responsible_person?: string
  alert_percentage?: number
  icon?: string
  color?: string
  sort_order?: number
  common_purchases?: CreateCommonPurchasePayload[]
}

/** POST /api/budgets */
export async function createBudget(body: CreateBudgetPayload): Promise<{ id: string }> {
  const res = await authenticatedRequest<ApiResponse<{ id: string }>>(
    '/api/budgets',
    { method: 'POST', body: JSON.stringify(body) },
  )
  return unwrapData('/api/budgets', res, 'Create budget')
}

export interface AppendCommonPurchasesPayload {
  purchases: CreateCommonPurchasePayload[]
}

/** POST /api/budgets/:id/common-purchases */
export async function appendBudgetCommonPurchases(
  budgetId: string,
  body: AppendCommonPurchasesPayload,
): Promise<void> {
  const res = await authenticatedRequest<ApiResponse<Record<string, unknown>>>(
    `/api/budgets/${encodeURIComponent(budgetId)}/common-purchases`,
    { method: 'POST', body: JSON.stringify(body) },
  )
  if (res.status !== 'success') {
    throw new Error(
      typeof res.error === 'string' ? res.error : 'Append presets failed',
    )
  }
}
